src/lib/homestore/heap_chunk_selector.cpp

```diff
@@ -0,0 +1,131 @@
+#include "heap_chunk_selector.h"
+
+#include <execution>
+#include <algorithm>
+#include <utility>
+
+#include <sisl/logging/logging.h>
+
+namespace homeobject {
+// https://github.com/eBay/HomeObject/pull/30#discussion_r1331112743
+// we make the following assumptions
+// 1 homestore will initialize HeapChunkSelector by adding all the chunks single threaded
+// 2 we do not need dynamic chunk requirements
+
+// it means after the single thread initialization,
+// 1 the key collection of m_per_dev_heap will never change.
+// 2 the key collection of m_chunks will never change
+void HeapChunkSelector::add_chunk(csharedChunk& chunk) {
+    VChunk vchunk(chunk);
+    auto pdevID = vchunk.get_pdev_id();
+    // add this find here, since we don`t want to call make_shared in try_emplace every time.
+    auto it = m_per_dev_heap.find(pdevID);
+    if (it == m_per_dev_heap.end()) it = m_per_dev_heap.emplace(pdevID, std::make_shared< PerDevHeap >()).first;
+    auto& avalableBlkCounter = it->second->available_blk_count;
+    avalableBlkCounter.fetch_add(vchunk.available_blks());
+
+    m_chunks.emplace(vchunk.get_chunk_id(), chunk);
+
+    auto& heapLock = it->second->mtx;
+    auto& heap = it->second->m_heap;
+    std::lock_guard< std::mutex > l(heapLock);
+    heap.emplace(chunk);
+}
+
+// select_chunk will only be called in homestore when creating a shard.
+csharedChunk HeapChunkSelector::select_chunk(homestore::blk_count_t count, const homestore::blk_alloc_hints& hint) {
+    auto& chunkIdHint = hint.chunk_id_hint;
```

Comment:
chunk_id_hint or chunkIdHint is not used anywhere in this function, what is the reason of checking it here? 

There is no use case consumer will call select_chunk with a chunk_id_hint, right?

The only use case I can see is remove `mark_chunk_selected`, replace by calling `select_chunk` too with a provided `chunk_id_hint`; But I also don't see why `mark_chunk_selected` is needed during recovery boot. 

Ideally HeapChunkSelector should wait to build heap until ShardManager completes its recovery, so that at that point, ShardManager knows which are the chunks that are with open_shards and those chunks should not participate heap building but only be added to m_chunks. It is okay to do this with another PR.

Reply:
I think Yaojie probabaly right.  The chunk information recovery is happending in HS layer, `PhysicalDev::load_chunks` will call `chunk_found_cb` and I believe the chunk should be added into selector here (i.e in the lambda of  DeviceManager::load_vdevs() )

Later when HO recover its shards, for every open shards it recovered , it marks the chunk as selected from selector.

Shard/Chunk are from different layer so if we want to wait ShardManager recover completes, we need to let HO get all chunks belongs to a vdev, which,  IMO is breaking the isolation

Reply:
ShardManager only need to inform HeapChunkSelector with  a list of "chunk ids" (which has open shards) and at that point, HeapChunkSelector can start building its heap by excluding those chunks with open shards on. 
ShardManager is already recovering from metablks and it doesn't need to know anything about chunks/vdev, just chunk ids is enough, right?

Reply:
ShardMgr has a list of open shards for sure but doesnt necessary has a list of "free shard" right?

Reply:
when homeobject restarts, it should first initialize homestore, because it need to load some metadata (for example, shard >chunk mapping) from metaservice. and then it can intialize shardmanager, and now shardmanager is able to know which shards are open. when intializing homestore, all the chunks will be added to ChunkSelector when initializing vdev.  
so we can not do this `m_shardMgr_inst->m_chunk_selector->start_recovery(open_chunk_ids);`. what we can do is only add all the chunks , and then get what shards are open ,and then mark them selected.

Reply:
i get your point @yamingk , i thinks the only key difference between our ideas is when should HeapChunkSelector build the per Pdev heap.

My implementation is buiding per pdev heap when adding chunk.
your idea is not building per pdev heap when adding chunk, but building the heaps after shardManager is recovered, so the HeapChunkSelector can build the heaps with the opened-shard-chunk information told by shardManager to exclude them from the heap.

logically, they can both achieve the goal. but `add_chunk` is also called by `release_chunk`(seal shard), so if using your idea, we need to add additional logic in `add_Chunk` to identify whether this add is a `release_chunk` call or just a `normal add chunk` when initializing, which may let the logic in this function a little confusing and complex.

so i prefer my current implementation

Reply:
@JacksonYao287 are you going to remove `chunk_id_hint` check  and keep `mark_chunk_selected` API or keep it and do consume it in `select_chunk` and retire `mark_chunk_selected` api? Either way if fine. you can make the call.

Reply:
@JacksonYao287 They can both achieve the goal. Difference is more efficient during bootup, e.g. we don't need to remove every elements from the top, put them into a tmp vector until we found a match for the chunkid being marked as selected, them add all the chunk in the tmp vector back to the heap. 

Let's say we have 64K chunks, and 200 open chunks, during boot up, this single mark chunks selected operation will do 200 * 64K * Log (64K), and these operations can all be zeroed, if we build heap after ShardManager recovery completes.

Reply:
> Difference is more efficient during bootup
agree, it is more efficient to build per dev heap after we know all the selected chunk.

>are you going to remove chunk_id_hint check
we can , but after that it is the reposibility of homestore to make sure using selector correctly. since it will not waste a lot efficiency , we could keep it like this as a sanity check.

Reply:
@yamingk i have make the change， pls take a look